### PR TITLE
[collectd 6] processes plugin: Migrate to collectd 6 data structures.

### PR DESCRIPTION
**Pull request by octo**
ChangeLog: processes plugin: The metric schema has been aligned with OpenTelemetry semantic conventions.

### CODE REVIEW

**Comment by eero-t** [comment]
Btw. do the open issues and PRs related to processes plugin have anything that should be taken into account here?
* Short-lived process handling:
  - https://github.com/collectd/collectd/issues/3026
  - https://github.com/collectd/collectd/pull/3027
* https://github.com/collectd/collectd/issues/2846
* https://github.com/collectd/collectd/pull/4122
* https://github.com/collectd/collectd/pull/1157

File: src/processes.c [hunk 1]
```diff
@@ -1424,8 +1600,8 @@ static int ps_read_process(long pid, process_entry_t *ps, char *state) {
     return 0;
   }
 
-  cpu_user_counter = atoll(fields[11]);
-  cpu_system_counter = atoll(fields[12]);
+  long long cpu_user_counter = atoll(fields[11]);
+  long long cpu_system_counter = atoll(fields[12]);
```
Comment (eero-t) [inline src/processes.c:1604]: line 1576 could have comment like:
`// field numbers are ones listed for 'stat' file in 'man proc' - 3`

**Review by eero-t — CHANGES_REQUESTED** [review]
Hard to review as this is so huge, IMHO it would have been better as split into multiple PRs. I think I noticed couple of bugs though.

PS. following PRs could do further cleanups:
*  split FreeBSD/NetBSD/OpenBSD and AIX cmdline parsing into a separate functions, similarly to how it's done for Linux & Solaris
* Order `process_count[<state>] = 0` statements and switch cases alphabetically so that it's easier to check them against each other

(Not doing them in this PR, or rebasing things until merging, helps the review.)

File: src/processes.c [hunk 2]
```diff
@@ -1983,8 +2048,10 @@ static int ps_read(void) {
         pse.vmem_minflt_counter = task_events_info.cow_faults;
         pse.vmem_majflt_counter = task_events_info.faults;
 
-        pse.cpu_user_counter = task_absolutetime_info.total_user;
-        pse.cpu_system_counter = task_absolutetime_info.total_system;
+        pse.cpu_user_counter =
+            ((fpcounter_t)task_absolutetime_info.total_user) / 1e9;
+        pse.cpu_system_counter =
+            ((fpcounter_t)task_absolutetime_info.total_system) / 1e9;
```
[the remark this answers is not in the document]
Comment (octo) [reply src/processes.c:2054]: I think the original code was incorrect: according to https://codereview.chromium.org/16818003 the uint64_t contain nanoseconds. Unfortunately Apple's documentation doesn't say anything about the unit of these counter :(

File: src/processes.c [hunk 3]
```diff
@@ -2086,129 +2152,119 @@ static int ps_read(void) {
...
-                ps_get_cmdline(pid, pse.name, cmdline, sizeof(cmdline)), &pse);
+    ps_list_add(pse.name, &pse);
   }
 
   closedir(proc);
 
-  if (read_file_contents("/proc/stat", buffer, sizeof(buffer) - 1) <= 0) {
-    ERROR("Cannot read `/proc/stat`");
+  char buffer[4096] = {0};
+  ssize_t n = read_text_file_contents("/proc/stat", buffer, sizeof(buffer) - 1);
```
[the remark this answers is not in the document]
Comment (octo) [reply src/processes.c:2226]: ✅ done

File: src/processes.c [hunk 4]
```diff
@@ -2086,129 +2152,119 @@ static int ps_read(void) {
...
+   * occurences of 'R' as character indicating the running state is typically
+   * zero. Due to processes are actually changing state during the evaluation
+   * of it's stat(s). The 'procs_running' number in /proc/stat on the other
+   * hand is more accurate, and can be retrieved in a single 'read' call. */
+  process_count[STATE_RUNNING] = (gauge_t)procs_running(buffer);
 
-  for (procstat_t *ps_ptr = list_head_g; ps_ptr != NULL; ps_ptr = ps_ptr->next)
-    ps_submit_proc_list(ps_ptr);
+  for (procstat_t *ps = list_head_g; ps != NULL; ps = ps->next)
+    ps_submit_proc_list(ps);
```
[the remark this answers is not in the document]
Comment (octo) [reply src/processes.c:2242]: ✅ done

File: src/processes.c [hunk 5]
```diff
@@ -2631,146 +2676,137 @@ static int ps_read(void) {
...
       }
 
-      memset(&pse, 0, sizeof(pse));
-
-      pse.id = procentry[i].pi_pid;
-      pse.num_lwp = procentry[i].pi_thcount;
-      pse.num_proc = 1;
+      process_entry_t pse = {
+          .id = procentry[i].pi_pid,
+          .command_line = cargs,
```
[the remark this answers is not in the document]
Comment (octo) [reply src/processes.c:2759]: Wow, good catch!

✅ done

**Review by octo — COMMENTED** [review]
Thanks for your review @eero-t! I realize this PR is way too large. I should have started refactoring the code as it was very convoluted and it's still more complicated than I'd like. Alas, by the time I realized the extend I was into too deep 🫤

File: src/processes.c [hunk 6]
```diff
@@ -441,223 +455,252 @@ static procstat_t *ps_list_register(const char *name, const char *regexp) {
...
+  snprintf(file, sizeof(file), "/proc/%ld/cmdline", pid);
+  ssize_t status = read_text_file_contents(file, buf, buf_len);
+  if (status < 0) {
+    ERROR("process plugin: Reading \"%s\" failed: %s", file, STRERRNO);
+    return (int)status;
+  }
+  if (status == 0) {
+    /* cmdline not available; e.g. kernel thread, zombie */
+    if (NULL == name) {
+      return EINVAL;
```
[the remark this answers is not in the document]
Comment (eero-t) [reply src/processes.c:577]: Ok, let's leave this as reminder for that potential future PR.

**Review by eero-t — APPROVED** [review]
Checked the earlier commented items and new commits (I assume rebase did not change earlier ones).   Review is somewhat superficial due to the huge amount of changes, but that's best I could do => approved.

**Comment by octo** [comment]
Thanks @eero-t!

**Pull request merged**